File: JSCheatSheet/Graphs/Problem2.js
class Problem2{

  bfs(i, j, vis, grid){
    vis[i][j] = 1;
    const n = grid.length;
    const m = grid[0].length;

    // const queue = []
    // const enqueue = (item) => {
    //   queue.push(item);
    // }
    // const dequeue = () => {
    //   queue.shift();
    // }
    // enqueue([i,j]);
    const queue = [[i,j]];

    while(queue.length>0){
      const [r,c] = queue.shift();
      for(let dr=-1; dr<=1; dr++){
        for(let dc=-1; dc<=1; dc++){
          const newr = dr + r;
          const newc = dc + c;
          if(0<=newr && newr<n && 0 <= newc && newc<m && grid[newr][newc] == "1" && !vis[newr][newc]){
            queue.push([newr, newc]);
            vis[newr][newc] = 1;
          }
        }
      }
    }
  }

  numIslands(grid){
    const n = grid.length;
    const m = grid[0].length;
    const vis = Array.from({length: n}, () => Array(m).fill(0));
    let cnt = 0;

    for(let i=0; i<n; i++){
      for(let j=0; j<m; j++){
        if(grid[i][j]=="1" && vis[i][j]==0){
          cnt += 1;
          this.bfs(i,j,vis,grid);
        }
      }
    }
    return cnt;
  }
}


const grid = [
    [0, 1, 1, 1, 0, 0, 0],
    [0, 0, 1, 1, 0, 1, 0]
];

const obj = new Problem2();
const result = obj.numIslands(grid);
console.log("number of islands:", result);

File: C++/Graphs/Problem2.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;
//Number of islands I
class Problem2{
  public:

    int bfs(int i, int j, vector<vector<int>> &vis, vector<vector<char>> &grid){
      int n = grid.size();
      int m = grid[0].size();
      vis[i][j] = 1;
      queue<pair<int,int>> q;
      q.push({i, j});
      while(!q.empty()){
        int r = q.front().first;
        int c = q.front().second;
        q.pop();

        for(int dr=-1; dr<=1; dr++){
          for(int dc=-1; dc<=1; dc++){
            int newr = r + dr;
            int newc = c + dc;
            if(newr>=0 && newr<n && newc>=0 && newc<m && grid[newr][newc]=='1' && !vis[newr][newc]){
                vis[newr][newc] = 1;
                q.push({newr, newc});
            }
          }
        }
      }
    }

    int numIslands(vector<vector<char>> &grid){
      int n = grid.size();
      int m = grid[0].size();
      vector<vector<int>> vis(n, vector<int>(m, 0));
      int cnt = 0;
      for(int i=0; i<n; i++){
        for(int j=0; j<m; j++){
          if(grid[i][j] == '1' && !vis[i][j]){
            cnt++;
            bfs(i, j, vis, grid);
          }
        }
      }
      return cnt;
    }
};

int main(){

  vector<vector<char>> grid = {
    {'0','1','1','1','0','0','0'},
    {'0','0','1','1','0','1','0'}
  };

  Problem2 obj;
  int result = obj.numIslands(grid);
  cout << "number of islands: " << result;

  return 0;
}
